runtime/catalog-karapace/src/main/java/io/aklivity/zilla/runtime/catalog/karapace/internal/KarapaceCatalogHandler.java

```diff
@@ -119,26 +137,80 @@ public int resolve(
+            schemaId = schemaIds.get(schemaKey).id;
         }
         else
         {
-            String response = sendHttpRequest(MessageFormat.format(SUBJECT_VERSION_PATH, subject, version));
-            schemaId = response != null ? request.resolveResponse(response) : NO_SCHEMA_ID;
-            if (schemaId != NO_SCHEMA_ID)
+            CompletableFuture<CachedSchemaId> future = cachedSchemaIds.get(schemaKey);
+            if (future == null)
             {
-                schemaIds.put(checkSum, new CachedSchemaId(System.currentTimeMillis(), schemaId));
+                CompletableFuture<CachedSchemaId> newFuture = new CompletableFuture<>();
+                future = cachedSchemaIds.putIfAbsent(schemaKey, newFuture);
+                if (future == null)
+                {
+                    newFuture = CompletableFuture.supplyAsync(() ->
+                    {
+                        String response = sendHttpRequest(MessageFormat.format(SUBJECT_VERSION_PATH, subject, version));
+                        return new CachedSchemaId(System.currentTimeMillis(),
+                            response != null ? request.resolveResponse(response) : NO_SCHEMA_ID);
+                    });
+                    future = newFuture;
+                }
             }
-            else if (schemaIds.containsKey(checkSum))
+            assert future != null;
+            try
             {
-                schemaId = schemaIds.get(checkSum).id;
+                CachedSchemaId cachedSchemaId = future.get();
+                schemaId = cachedSchemaId.id;
+                if (schemaId != NO_SCHEMA_ID && (System.currentTimeMillis() - cachedSchemaId.timestamp) < maxAgeMillis)
+                {
+                    schemaIds.put(schemaKey, cachedSchemaId);
+                }
+                else
+                {
+                    CompletableFuture<CachedSchemaId> newFuture =
+                        cachedSchemaIds.computeIfPresent(schemaKey, (key, existingFuture) ->
+                        {
+                            if (existingFuture.isDone() || existingFuture.isCompletedExceptionally())
+                            {
+                                CompletableFuture<CachedSchemaId> id = CompletableFuture.supplyAsync(() ->
+                                {
+                                    String response = sendHttpRequest(MessageFormat.format(SUBJECT_VERSION_PATH,
+                                        subject, version));
+                                    return new CachedSchemaId(System.currentTimeMillis(),
+                                        response != null ? request.resolveResponse(response) : NO_SCHEMA_ID);
+                                });
+                                return id;
+                            }
+                            else
+                            {
+                                return existingFuture;
+                            }
+                        });
```

Comment:
I think we can perhaps simplify the logic by creating the candidate newFuture and calling:
```
CompletableFuture<> newFuture = new CompletableFuture<CachedSchemaId>();
CompletableFuture<> future = cachedSchemaIds.merge(schemaKey, newFuture, (v1, v2) -> v1 != null && !v1.isDone() ? v1 : v2);
if (future == newFuture)
{
    // send request
    // complete future
}
cachedSchemaId  = future.get();
```